File: dot_claude/hooks/vault-handoff-check.mjs
#!/usr/bin/env node
// Stop hook: enforce 'documentation is part of done'.
//
// If the session made meaningful changes (file edits or git commits) in a
// project workspace but never wrote a vault inbox handoff, block the stop once
// with a reminder. stop_hook_active guards against loops: the nudge fires at
// most once per stop, and the second stop always passes.
//
// Node (not python): devpod images have no system python, but mise ships node
// in every pod and on the host. Portable across host and devpods: paths are
// home-relative, devpod workspaces (/workspaces/<repo>) are watched, and the
// hook stays silent when the vault is not present. Fails open on any error.

import fs from "node:fs";
import os from "node:os";
import path from "node:path";

const HOME = os.homedir();
const VAULT_INBOX = path.join(HOME, "Projects", "claude-obsidian", "inbox");
const VAULT_INBOX_MARKER = "claude-obsidian/inbox";
const WATCHED = [path.join(HOME, "Projects"), path.join(HOME, "Claude"), "/workspaces"];
const EXCLUDED_DIR_PARTS = ["claude-obsidian", "vault-obsidian"];
const EDIT_TOOLS = new Set(["Edit", "Write", "MultiEdit", "NotebookEdit"]);
const MIN_EDITS = 2;

function realpathSafe(p) {
  try {
    return fs.realpathSync(p);
  } catch {
    return p;
  }
}

function inWatched(cwd) {
  const real = realpathSafe(cwd);
  for (const prefix of WATCHED) {
    for (const p of [prefix, realpathSafe(prefix)]) {
      if (real === p || real.startsWith(p + path.sep) || cwd.startsWith(p + path.sep)) {
        return true;
      }
    }
  }
  return false;
}

function main() {
  const data = JSON.parse(fs.readFileSync(0, "utf8"));
  if (data.stop_hook_active) return;
  if (!fs.existsSync(VAULT_INBOX)) return;

  const cwd = data.cwd || "";
  if (!cwd || !inWatched(cwd)) return;
  if (EXCLUDED_DIR_PARTS.some((part) => cwd.includes(part))) return;

  const transcriptPath = data.transcript_path || "";
  if (!transcriptPath || !fs.existsSync(transcriptPath)) return;

  let edits = 0;
  let committed = false;
  let handoffWritten = false;
  for (const line of fs.readFileSync(transcriptPath, "utf8").split("\n")) {
    let entry;
    try {
      entry = JSON.parse(line);
    } catch {
      continue;
    }
    const content = entry?.message?.content;
    if (!Array.isArray(content)) continue;
    for (const block of content) {
      if (!block || block.type !== "tool_use") continue;
      const name = block.name || "";
      const input = block.input || {};
      const filePath = String(input.file_path || "");
      if (filePath.includes(VAULT_INBOX_MARKER)) {
        handoffWritten = true;
      } else if (EDIT_TOOLS.has(name) && filePath) {
        edits += 1;
      } else if (name === "Bash") {
        const command = String(input.command || "");
        if (command.includes(VAULT_INBOX_MARKER)) {
          handoffWritten = true;
        } else if (command.includes("git commit")) {
          committed = true;
        }
      }
    }
  }

  const meaningful = committed || edits >= MIN_EDITS;
  if (meaningful && !handoffWritten) {
    process.stdout.write(JSON.stringify({
      decision: "block",
      reason:
        "This session made changes but no vault inbox handoff was written. " +
        "Per the Project Agent protocol, write a handoff to " +
        "~/Projects/claude-obsidian/inbox/<YYYY-MM-DD>-<slug>.md using the " +
        "template in the project CLAUDE.md (What Was Done, Key Decisions, " +
        "Gotchas, References), then commit and push the vault if working in " +
        "a devpod. If the changes were genuinely trivial (typo-level, no " +
        "knowledge worth keeping), state that in one line and stop.",
    }));
  }
}

try {
  main();
} catch {
  // fail open
}
process.exit(0);

File: dot_claude/hooks/vault-handoff-check.py
#!/usr/bin/env python3
"""Stop hook: enforce 'documentation is part of done'.

Design rule (Bitter Lesson): the hook supplies environmental facts and a single
nudge; the model judges whether a handoff is warranted. No transcript tool
parsing and no tool-name allowlists; those rot silently as the harness evolves.

Signals, all environmental:
- session start: timestamp of the first transcript line (the only harness-schema
  dependence; if unreadable, degrade to nudging, never to silence)
- handoff written: any vault inbox file modified since session start
- meaningful work (git repos only): dirty working tree or a commit since start

In a watched non-git directory, or when session start is unknown, the nudge
fires unconditionally once; a session with nothing to document says so in one
line and stops. stop_hook_active guarantees at most one nudge per stop. Fails
open on error. Python comes from mise in devpods and from the OS on the host.
"""
import json
import os
import subprocess
import sys
from datetime import datetime

HOME = os.path.expanduser("~")
VAULT_INBOX = os.path.join(HOME, "Projects", "claude-obsidian", "inbox")
WATCHED = [os.path.join(HOME, "Projects"), os.path.join(HOME, "Claude"), "/workspaces"]
EXCLUDED_DIR_PARTS = ("claude-obsidian", "vault-obsidian")
CLOCK_SLACK = 120


def in_watched(cwd: str) -> bool:
    real = os.path.realpath(cwd)
    for prefix in WATCHED:
        for p in (prefix, os.path.realpath(prefix)):
            if real == p or real.startswith(p + os.sep) or cwd.startswith(p + os.sep):
                return True
    return False


def session_start_epoch(transcript_path: str):
    if not transcript_path or not os.path.exists(transcript_path):
        return None
    try:
        with open(transcript_path, "r", errors="replace") as fh:
            for line in fh:
                try:
                    entry = json.loads(line)
                except json.JSONDecodeError:
                    continue
                ts = entry.get("timestamp")
                if isinstance(ts, str) and ts:
                    return datetime.fromisoformat(ts.replace("Z", "+00:00")).timestamp()
    except (OSError, ValueError):
        pass
    return None


def handoff_written_since(start: float) -> bool:
    try:
        for name in os.listdir(VAULT_INBOX):
            path = os.path.join(VAULT_INBOX, name)
            if os.path.isfile(path) and os.path.getmtime(path) >= start - CLOCK_SLACK:
                return True
    except OSError:
        pass
    return False


def git(args, cwd):
    return subprocess.run(
        ["git", *args], cwd=cwd, capture_output=True, text=True, timeout=10
    )


def git_activity(cwd: str, start: float):
    """True/False when cwd is a git repo, None when it is not (or git is absent)."""
    try:
        inside = git(["rev-parse", "--is-inside-work-tree"], cwd)
    except (OSError, subprocess.SubprocessError):
        return None
    if inside.returncode != 0 or inside.stdout.strip() != "true":
        return None
    try:
        status = git(["status", "--porcelain"], cwd)
        if status.returncode == 0 and status.stdout.strip():
            return True
        head_time = git(["log", "-1", "--format=%ct"], cwd)
        if head_time.returncode == 0 and head_time.stdout.strip().isdigit():
            return int(head_time.stdout.strip()) >= start - CLOCK_SLACK
    except (OSError, subprocess.SubprocessError):
        return None
    return False


def main() -> None:
    data = json.load(sys.stdin)
    if data.get("stop_hook_active"):
        return

    if not os.path.isdir(VAULT_INBOX):
        return

    cwd = data.get("cwd", "")
    if not cwd or not in_watched(cwd):
        return
    if any(part in cwd for part in EXCLUDED_DIR_PARTS):
        return

    start = session_start_epoch(data.get("transcript_path", ""))
    if start is not None:
        if handoff_written_since(start):
            return
        if git_activity(cwd, start) is False:
            return

    print(json.dumps({
        "decision": "block",
        "reason": (
            "Before stopping, apply the Project Agent protocol: if this session "
            "did meaningful work (built, configured, fixed, decided, or learned "
            "something) and no vault inbox handoff covers it yet, write one to "
            "~/Projects/claude-obsidian/inbox/<YYYY-MM-DD>-<slug>.md, then commit "
            "and push the vault if working in a devpod. If nothing worth "
            "documenting happened, say so in one line and stop."
        ),
    }))


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass
    sys.exit(0)
